# Flag and replace deprecated component usage (`replace-deprecated-component`)

This rule ...

## Rule Details

Examples of **incorrect** code for this rule:

```typescript
...
```

Examples of **correct** code for this rule:

```typescript
...
```

## When Not To Use It

...

## Auto-fixable?

Yes ✔️
